### ISSUE TITLE
add clojure.repl/doc

### ISSUE THREAD

**Issue by roman01la** [borkdude/sci#263]
Is it possible to have `clojure.repl/doc` in sci?

**Comment by borkdude** [comment]
Is this for babashka or for a project you're building with sci? Can you give some context?

**Comment by roman01la** [comment]
Using sci for Slack bot currently, `doc` would be a useful function to have

**Comment by borkdude** [comment]
@roman01la Currently the clojure.core functions in sci do not have :doc metadata, although that can be fixed. But there might be a very quick and easy workaround to get what you want for now. Is the repo public and can you send me the link so I can take a look? Depending on where you call sci from, the fix might look different.

**Comment by roman01la** [comment]
It's not public. The library used via Node.js API.

**Comment by borkdude** [comment]
@roman01la The quick workaround (before a proper fix lands in sci) I had in mind was something like:

``` clojure
user=> (def doc-map {'+ (:doc (meta #'+))})
#'user/doc-map
user=> doc-map
{+ "Returns the sum of nums. (+) returns 0. Does not auto-promote\n  longs, will throw on overflow. See also: +'"}
user=> (defn sci-doc [sym] (get doc-map sym))
#'user/sci-doc
user=> (require '[sci.core :as sci])
nil
user=> (defn sci-doc [_ _ sym] (get doc-map sym))
#'user/sci-doc
user=> (sci/eval-string "(clojure.repl/doc +)" {:namespaces {'clojure.repl {'doc (with-meta sci-doc {:sci/macro true})}}})
"Returns the sum of nums. (+) returns 0. Does not auto-promote\n  longs, will throw on overflow. See also: +'"
```

**Comment by roman01la** [comment]
Ah yeah, that makes sense. How that would look like in JavaScript when using from Node?

**Comment by borkdude** [comment]
@roman01la 

Something like this could work as a short-term solution:

```
const { evalString, toJS } = require('@borkdude/sci');

var withMeta = evalString("with-meta");
var docMacro = evalString(`(fn [_ _ sym]
\`(or
    (get {~'inc "increments number"} ~sym)
    (-> (resolve '~sym) meta :doc)))`);

var sciOpts = {namespaces: {"clojure.repl": {doc: withMeta(docMacro, evalString("{:sci/macro true}"))},}};

var incDoc = evalString("(clojure.repl/doc inc)", sciOpts);
console.log(incDoc);

var fooDoc = evalString("(defn foo \"foo docs\" []) (clojure.repl/doc foo)", sciOpts);
console.log(fooDoc);
```

```
$ node index.js
increments number
foo docs
```

With a little bit of hackery the doc-strings can be obtained from a script and then dumped into some JS.

**Comment by borkdude** [comment]
@roman01la The proper fix (in contrast to the hack above) would be something like the branch `copy-vars` which I just pushed. This copies the :doc metadata to a sci var. I have done this for the functions `clojure.core/boolean` and `clojure.core/inc`.

The result:

```
$ script/compile-js
$ node
Welcome to Node.js v13.7.0.
Type ".help" for more information.
> require("./out/min/sci.min")
> const { evalString } = require("./out/min/sci.min")
undefined
> evalString("(clojure.repl/doc inc)")
'Returns a number one greater than num.'
```

We could do this for all vars. I would have to measure differences in performance. Until now all the core bindings are not reified as vars but just references to the function values.

One other thing to think about is printing. Sci by default doesn't allow printing, unless you bind the *out* stream to something.

**Comment by borkdude** [comment]
Note: when reifying clojure.core functions into vars, tight loops take a small performance hit:

```
$ multitime -n5 -s0 ./sci "(loop [val 0 cnt 1000000] (if (pos? cnt) (recur (inc val) (dec cnt)) val))"
1000000
1000000
1000000
1000000
1000000
===> multitime results
1: ./sci "(loop [val 0 cnt 1000000] (if (pos? cnt) (recur (inc val) (dec cnt)) val))"
            Mean        Std.Dev.    Min         Median      Max
real        2.075       0.056       2.011       2.045       2.151
```

The mean before was around 1.5seconds. I'm not saying this is a reason not to pursue this avenue, just something worth knowing.

**Comment by borkdude** [comment]
I now reified all core vars in sci in the `copy-vars` branch. Note: this adds about 180kb to the minified JS.

**Comment by borkdude** [comment]
@roman01la 

With the latest push:

```
> evalString("(clojure.repl/doc inc)", {namespaces: {"clojure.core": {println: console.log}}})
Returns a number one greater than num.
undefined
> evalString("(with-out-str (clojure.repl/doc inc))")
'Returns a number one greater than num.\n'
```

Does this look like a good start to you? Maybe give it a bit of testing? We can add `:arglists` as well to get closer to the output of the original `clojure.repl/doc`.

**Comment by borkdude** [comment]
With the latest push:

```
$ node
> const { evalString } = require("./out/min/sci.min")
> evalString("(clojure.repl/doc inc)", {namespaces: {"clojure.core": {println: function(...xs) { var strs = xs.map(function(x) { return x.toString(); }); console.log(...strs); }}}});
-------------------------
clojure.core/inc
([x])
  Returns a number one greater than num.
```

**Comment by borkdude** [comment]
@roman01la I released 0.0.13-alpha.8 where this is now possible:

```
$ node
Welcome to Node.js v13.7.0.
Type ".help" for more information.
> const { evalString } = require("@borkdude/sci")
undefined
> var println = function(...xs) {
...   var strs = xs.map(x => x.toString()); console.log(...strs); };
undefined
> evalString("(clojure.repl/doc inc)", {namespaces: {"clojure.core": {println: println}}})
-------------------------
clojure.core/inc
([x])
  Returns a number one greater than num.
```
